File: javascript/examples/test-local.ts
/**
 * Test local du SDK JavaScript SahelPay
 * 
 * Usage:
 *   cd sdks/javascript
 *   npm install
 *   npm run build
 *   npx tsx examples/test-local.ts
 */

// Import depuis le build local
import SahelPay from '../src/index.js';

const API_KEY = 'sk_test_demo123'; // Clé de test
const BASE_URL = 'http://localhost:3000'; // Backend local

async function testSDK() {
  console.log('🚀 Test SDK JavaScript SahelPay\n');
  console.log('='.repeat(50));

  // Initialisation
  const sahelpay = new SahelPay({
    secretKey: API_KEY,
    environment: 'sandbox',
    baseUrl: BASE_URL,
  });

  console.log('✅ Client initialisé');
  console.log(`   Base URL: ${BASE_URL}`);
  console.log(`   Mode: sandbox\n`);

  // Test 1: Lister les transactions
  console.log('📋 Test 1: Lister les transactions...');
  try {
    const { payments } = await sahelpay.payments.list({ limit: 5 });
    console.log(`   ✅ ${payments?.length || 0} transactions trouvées\n`);
  } catch (error: any) {
    console.log(`   ❌ Erreur: ${error.message}\n`);
  }

  // Test 2: Lister les liens de paiement
  console.log('🔗 Test 2: Lister les liens de paiement...');
  try {
    const links = await sahelpay.paymentLinks.list();
    console.log(`   ✅ ${links?.length || 0} liens trouvés\n`);
  } catch (error: any) {
    console.log(`   ❌ Erreur: ${error.message}\n`);
  }

  // Test 3: Stats Payouts
  console.log('💰 Test 3: Statistiques payouts...');
  try {
    const stats = await sahelpay.payouts.stats();
    console.log(`   ✅ Total: ${stats?.total || 0}`);
    console.log(`   ✅ Complétés: ${stats?.completed || 0}`);
    console.log(`   ✅ Volume: ${stats?.total_volume || 0} FCFA\n`);
  } catch (error: any) {
    console.log(`   ❌ Erreur: ${error.message}\n`);
  }

  // Test 4: Initier un paiement (simulation)
  console.log('💳 Test 4: Initier un paiement...');
  try {
    const payment = await sahelpay.payments.create({
      amount: 1000,
      provider: 'ORANGE_MONEY',
      customer_phone: '+22370000000',
      description: 'Test SDK JavaScript',
    });
    console.log(`   ✅ Paiement créé: ${payment.reference_id}`);
    console.log(`   ✅ Status: ${payment.status}`);
    console.log(`   ✅ USSD: ${payment.ussd_code || 'N/A'}\n`);
  } catch (error: any) {
    console.log(`   ⚠️ Erreur attendue (pas de marchand): ${error.message}\n`);
  }

  // Test 5: Webhook verification
  console.log('🔒 Test 5: Vérification signature webhook...');
  const testPayload = '{"event":"payment.success","data":{"amount":1000}}';
  const testSecret = 'whsec_test123';
  try {
    // Générer une signature valide
    const crypto = require('crypto');
    const validSig = crypto.createHmac('sha256', testSecret).update(testPayload).digest('hex');
    
    const isValid = sahelpay.webhooks.verifySignature(testPayload, validSig, testSecret);
    console.log(`   ✅ Signature valide: ${isValid}`);
    
    const isInvalid = sahelpay.webhooks.verifySignature(testPayload, 'bad_signature', testSecret);
    console.log(`   ✅ Signature invalide rejetée: ${!isInvalid}\n`);
  } catch (error: any) {
    console.log(`   ❌ Erreur: ${error.message}\n`);
  }

  console.log('='.repeat(50));
  console.log('✅ Tests terminés!\n');
}

testSDK().catch(console.error);

File: python/examples/test_local.py
#!/usr/bin/env python3
"""
Test local du SDK Python SahelPay

Usage:
    cd sdks/python
    pip install -e .
    python examples/test_local.py
"""

import sys
import os

# Ajouter le chemin du SDK local
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from sahelpay.client import Client
from sahelpay.exceptions import SahelPayError, APIError
import hmac
import hashlib

# Clé secrète du marchand SahelPay Demo (déchiffrée)
API_KEY = 'sk_test_SAHEL_123'
BASE_URL = 'http://localhost:3005'  # Backend Docker port


def test_sdk():
    print('🚀 Test SDK Python SahelPay\n')
    print('=' * 50)

    # Initialisation
    client = Client(
        secret_key=API_KEY,
        environment='sandbox',
        base_url=BASE_URL
    )

    print('✅ Client initialisé')
    print(f'   Base URL: {BASE_URL}')
    print('   Mode: sandbox\n')

    # Test 1: Lister les transactions
    print('📋 Test 1: Lister les transactions...')
    try:
        result = client.payments.list(limit=5)
        payments = result.get('payments', [])
        print(f'   ✅ {len(payments)} transactions trouvées\n')
    except SahelPayError as e:
        print(f'   ❌ Erreur: {e}\n')

    # Test 2: Lister les liens de paiement
    print('🔗 Test 2: Lister les liens de paiement...')
    try:
        links = client.payment_links.list()
        print(f'   ✅ {len(links)} liens trouvés\n')
    except SahelPayError as e:
        print(f'   ❌ Erreur: {e}\n')

    # Test 3: Stats Payouts
    print('💰 Test 3: Statistiques payouts...')
    try:
        stats = client.payouts.stats()
        print(f'   ✅ Total: {stats.total}')
        print(f'   ✅ Complétés: {stats.completed}')
        print(f'   ✅ Volume: {stats.total_volume} FCFA\n')
    except SahelPayError as e:
        print(f'   ❌ Erreur: {e}\n')

    # Test 4: Initier un paiement (simulation)
    print('💳 Test 4: Initier un paiement...')
    try:
        payment = client.payments.create(
            amount=1000,
            provider='ORANGE_MONEY',
            customer_phone='+22370000000',
            description='Test SDK Python',
            sandbox=True,
        )
        print(f'   ✅ Paiement créé: {payment.reference_id}')
        print(f'   ✅ Status: {payment.status}')
        print(f'   ✅ USSD: {payment.ussd_code or "N/A"}\n')
    except SahelPayError as e:
        print(f'   ⚠️ Erreur attendue (pas de marchand): {e}\n')

    # Test 5: Webhook verification
    print('🔒 Test 5: Vérification signature webhook...')
    test_payload = '{"event":"payment.success","data":{"amount":1000}}'
    test_secret = 'whsec_test123'
    try:
        # Générer une signature valide
        import time
        timestamp = str(int(time.time()))
        valid_sig = hmac.new(
            test_secret.encode(),
            f'{timestamp}.{test_payload}'.encode(),
            hashlib.sha256
        ).hexdigest()
        valid_header = f't={timestamp},v1={valid_sig}'

        is_valid = client.webhooks.verify_signature(test_payload, valid_header, test_secret)
        print(f'   ✅ Signature valide: {is_valid}')

        try:
            client.webhooks.verify_signature(test_payload, 't=123,v1=bad_signature', test_secret)
            print('   ❌ Signature invalide acceptée\n')
        except Exception:
            print('   ✅ Signature invalide rejetée\n')
    except Exception as e:
        print(f'   ❌ Erreur: {e}\n')

    print('=' * 50)
    print('✅ Tests terminés!\n')


if __name__ == '__main__':
    test_sdk()
